
| <h1> Assignment 5.2 </h1> | <h1> Build and Apply Multilayer Perceptron </h1> |
|--- | --- |
Name: Buan, Danilo Jr. G. <br>
Course and Section: CPE019/CPE32S3 <br>
Date Submitted: 03/26/24 <br>
Instructor: Engr. Roman M. Richard
Date Performed: 03/26/24


#Chosen Dataset and the Problem
- For this activity, I will be using the onlinefoods.csv from the Obesity Risk Dataset from Kaggle. In creating the model, the problem I am trying to solve is to predict the 'Output' variable based on the other features available in the dataset.


Kaggle Link: https://www.kaggle.com/datasets/sudarshan24byte/online-food-dataset


#Data Preprocessing

In [2]:
#Importing necessary libraries and the dataset
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, LabelEncoder, OneHotEncoder

df = pd.read_csv("onlinefoods.csv")

In [3]:
df.info()

df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 388 entries, 0 to 387
Data columns (total 13 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Age                         388 non-null    int64  
 1   Gender                      388 non-null    object 
 2   Marital Status              388 non-null    object 
 3   Occupation                  388 non-null    object 
 4   Monthly Income              388 non-null    object 
 5   Educational Qualifications  388 non-null    object 
 6   Family size                 388 non-null    int64  
 7   latitude                    388 non-null    float64
 8   longitude                   388 non-null    float64
 9   Pin code                    388 non-null    int64  
 10  Output                      388 non-null    object 
 11  Feedback                    388 non-null    object 
 12  Unnamed: 12                 388 non-null    object 
dtypes: float64(2), int64(3), object(8)


Age  Gender Marital Status Occupation  Monthly Income  \
0   20  Female         Single    Student       No Income   
1   24  Female         Single    Student  Below Rs.10000   
2   22    Male         Single    Student  Below Rs.10000   
3   22  Female         Single    Student       No Income   
4   22    Male         Single    Student  Below Rs.10000   

  Educational Qualifications  Family size  latitude  longitude  Pin code  \
0              Post Graduate            4   12.9766    77.5993    560001   
1                   Graduate            3   12.9770    77.5773    560009   
2              Post Graduate            3   12.9551    77.6593    560017   
3                   Graduate            6   12.9473    77.5616    560019   
4              Post Graduate            4   12.9850    77.5533    560010   

  Output   Feedback Unnamed: 12  
0    Yes   Positive         Yes  
1    Yes   Positive         Yes  
2    Yes  Negative          Yes  
3    Yes   Positive         Yes  
4    Yes   Positive         Yes

In [4]:
# Define categorical columns to be label-encoded
categorical_cols = ['Gender', 'Marital Status', 'Occupation', 'Monthly Income', 'Educational Qualifications', 'Feedback', 'Output']

label_encoder = LabelEncoder()

for col in categorical_cols:
    df[col] = label_encoder.fit_transform(df[col])

# Check the updated DataFrame
df.head()

Age  Gender  Marital Status  Occupation  Monthly Income  \
0   20       0               2           3               4   
1   24       0               2           3               2   
2   22       1               2           3               2   
3   22       0               2           3               4   
4   22       1               2           3               2   

   Educational Qualifications  Family size  latitude  longitude  Pin code  \
0                           2            4   12.9766    77.5993    560001   
1                           0            3   12.9770    77.5773    560009   
2                           2            3   12.9551    77.6593    560017   
3                           0            6   12.9473    77.5616    560019   
4                           2            4   12.9850    77.5533    560010   

   Output  Feedback Unnamed: 12  
0       1         1         Yes  
1       1         1         Yes  
2       1         0         Yes  
3       1         1         Yes  
4       1         1         Yes

In [5]:
#Dropping unnecessary columns
df = df.drop(columns = ["Unnamed: 12"])

df.head()

Age  Gender  Marital Status  Occupation  Monthly Income  \
0   20       0               2           3               4   
1   24       0               2           3               2   
2   22       1               2           3               2   
3   22       0               2           3               4   
4   22       1               2           3               2   

   Educational Qualifications  Family size  latitude  longitude  Pin code  \
0                           2            4   12.9766    77.5993    560001   
1                           0            3   12.9770    77.5773    560009   
2                           2            3   12.9551    77.6593    560017   
3                           0            6   12.9473    77.5616    560019   
4                           2            4   12.9850    77.5533    560010   

   Output  Feedback  
0       1         1  
1       1         1  
2       1         0  
3       1         1  
4       1         1

In [6]:
#Splitting the data into training and testing datasets
X = df.drop(columns=['Output'])
y = df['Output']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [7]:
# Display the structure of the dataset
print("Feature matrix:", X_train.shape)
print("Target matrix:", X_test.shape)
print("Feature matrix:", y_train.shape)
print("Target matrix:", y_test.shape)

Feature matrix: (310, 11)
Target matrix: (78, 11)
Feature matrix: (310,)
Target matrix: (78,)


#Creating my own model

In [8]:
import tensorflow as tf
from tensorflow import keras

# Define input layer
inputs = tf.keras.layers.Input(shape=(11,))

# Add dense hidden layers with L2 regularization
x = tf.keras.layers.Dense(128, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(0.01))(inputs)
x = tf.keras.layers.Dropout(0.5)(x)

# Add more dense hidden layers with L2 regularization
x = tf.keras.layers.Dense(64, activation='relu', kernel_regularizer=tf.keras.regularizers.l2(0.01))(x)
x = tf.keras.layers.Dropout(0.5)(x)

# Output layer with softmax activation (assuming 5 classes)
outputs = tf.keras.layers.Dense(5, activation='softmax')(x)

# Create the model
model = tf.keras.Model(inputs=inputs, outputs=outputs)

# Compile the model with a lower learning rate
model.compile(optimizer=tf.keras.optimizers.Adam(lr=0.0001),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


In [18]:
model.fit(X_train, y_train, epochs=15, batch_size = 64, validation_split = 0.2)

Epoch 1/15
4/4 [==============================] - 0s 39ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 2/15
4/4 [==============================] - 0s 19ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 3/15
4/4 [==============================] - 0s 13ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 4/15
4/4 [==============================] - 0s 12ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 5/15
4/4 [==============================] - 0s 18ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 6/15
4/4 [==============================] - 0s 13ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 7/15
4/4 [==============================] - 0s 21ms/step - loss: 0.5533 - accuracy: 0.7581 - val_loss: 0.5533 - val_accuracy: 0.7581
Epoch 8/15
4/4 [===========

#Evaluating the accuracy of my model

In [19]:
# Evaluate the accuracy of the model
test_loss, test_acc = model.evaluate(X_test, y_test, verbose=1)
print('Test loss:', test_loss)
print('Test accuracy:', test_acc)

3/3 [==============================] - 0s 5ms/step - loss: 0.4528 - accuracy: 0.8462
Test loss: 0.4527682363986969
Test accuracy: 0.8461538553237915


#Conclusion, Observation and Analysis

**Conclusion**
- After performing this activity, I have concluded that the neural network trained the 'onlinefoods.csv' dataset shows an acceptable predictive ability with an accuracy of 84.615%. This indicates that the model is effective in learning the provided features to make predictions on unseen data. The preprocessing of data also helped in building a model. Although this can still be further improved to maximize the model's performance for real-world applications.

**Observation**
- Upon evaluating the model I trained, I noticed a few interesting things. First off, the preprocessing steps, like turning categorical variables into numbers, really helped get the data ready for training. It made things easier for the model to understand. Secondly, when I tested the model, it seemed to do pretty well on new data. This tells me that it's not just memorizing the training data but actually learning underlying patterns. Still, I need to remember that the model might not perform the same way on different datasets or tougher problems. Even though the accuracy looks good, I know there's more to the story, so I'll need to dig deeper and look at other factors to fully understand how well the model is doing.

**Analysis**
- After analyzing the numerous steps I have done in creating the model for this dataset, there is still some important areas to explore further. The model's architecture, like the number of layers and units, was crucial for its performance. I think trying different setups could improve accuracy. Also, adjusting learning rates and dropout rates could help. Checking for biases or imbalances in the data is essential too. Overall, continuous analysis and adjustments are vital for creating effective machine learning models.